rplidar.go

```diff
@@ -98,70 +104,108 @@ func newRplidar(ctx context.Context, _ resource.Dependencies, c resource.Config,
-
-	if rp.close {
-		return nil, errors.New("resource (rplidar) is closed")
+// setupRPLiDAR starts the motor, in necessary, warms up the device, and ensures data returned to the
+// user is valid.
+func (rp *Rplidar) setupRPLidar(ctx context.Context) error {
+	// Start the motor
+	// Note: S1 rplidars do not need to start the motor before scanning can begin
+	if rplidarModelByteMap[rp.device.model] != "S1" {
+		rp.logger.Debug("starting motor")
+		rp.device.driver.StartMotor()
 	}
 
-	if !rp.started {
-		return nil, errors.New("resource (rplidar) failed to initialize properly")
-	}
+	// Setup rplidar scan and scan once as per warmup procedure
+	rp.device.driver.StartScan(false, true)
+	rp.nodes = gen.New_measurementNodeHqArray(defaultNodeSize)
 
-	pc, err := rp.getPointCloud(ctx)
-	if err != nil {
-		return nil, err
+	goutils.SelectContextOrWait(ctx, time.Duration(defaultWarmupNumDiscardedScans)*time.Second)
+	if _, err := rp.scan(ctx, defaultWarmupNumDiscardedScans); err != nil {
+		return err
 	}
-	return pc, nil
+
+	return nil
 }
 
-// Images is a part of the camera interface but is not implemented for the rplidar.
-func (rp *Rplidar) Images(ctx context.Context) ([]camera.NamedImage, resource.ResponseMetadata, error) {
-	return nil, resource.ResponseMetadata{}, errors.New("images unimplemented")
+// cachePointCloudLoop is a background process that repeatedly gets point cloud data from the rplidar
+// and caches it for later access. This will reduce delays in returning data and prevent overcrowding
+// of the rplidar's serial line.
+func (rp *Rplidar) cachePointCloudLoop(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			pc, err := rp.scan(ctx, defaultNumScans)
+			if err != nil {
+				rp.logger.Debugf("issue getting pointcloud to cache: %v", err)
+			}
+
+			rp.cacheMutex.Lock()
+			rp.cachedPointCloud = pc
+			rp.cacheMutex.Lock()
+		}
+	}
 }
 
+// scan uses the serial connection to the rplidar to get data and creates a pointcloud from the
+// returned distances and angles
 func (rp *Rplidar) scan(ctx context.Context, numScans int) (pointcloud.PointCloud, error) {
+	rp.deviceMutex.Lock()
+	defer rp.deviceMutex.Unlock()
+
 	pc := pointcloud.New()
-	nodeCount := int64(rp.nodeSize)
 
 	var dropCount int
+	nodeCount := int64(defaultNodeSize)
 	for i := 0; i < numScans; i++ {
-		nodeCount = int64(rp.nodeSize)
-		result := rp.device.driver.GrabScanDataHq(rp.nodes, &nodeCount, defaultTimeout)
+		result := rp.device.driver.GrabScanDataHq(rp.nodes, &nodeCount, defaultTimeoutMs)
 		if Result(result) != ResultOk {
 			return nil, fmt.Errorf("bad scan: %w", Result(result).Failed())
 		}
 		rp.device.driver.AscendScanData(rp.nodes, nodeCount)
 
 		for pos := 0; pos < int(nodeCount); pos++ {
-			node := gen.MeasurementNodeHqArray_getitem(rp.nodes, pos)
+
+			var node gen.Rplidar_response_measurement_node_hq_t
+			if !rp.testing {
+				node = gen.MeasurementNodeHqArray_getitem(rp.nodes, pos)
+			} else {
```

Comment:
i think doing 1 or 2 would be fine for now. I prefer 1 since the only success case is returning an empty pointcloud, but theres no reason to change the structure you setup outside of that. This PR is just standing up tests, we can probably figure out how to mock the lidar some other time.